File: algorithms/easy/smallest_stable_index_1.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        N = len(nums)
        lpfx = [-1 for _ in range(N)]
        for i in range(N):
            lpfx[i] = nums[i] if i == 0 else max(lpfx[i-1], nums[i])
        rpfx = [-1 for _ in range(N)]
        for i in range(N-1, -1, -1):
            rpfx[i] = nums[i] if i == N-1 else min(nums[i], rpfx[i+1])
        for i in range(N):
            if lpfx[i] - rpfx[i] <= k:
                return i
        return -1

# Main section
for nums, k in [
                  ([5,0,1,4], 3),
                  ([3,2,1], 1),
                  ([0], 0),
               ]:
    print(f'nums, k = {nums}, {k}')
    sol = Solution()
    r = sol.firstStableIndex(nums, k)
    print(f'r = {r}')
    print('==================================')

File: algorithms/easy/smallest_stable_index_1.go
package main

import "fmt"

func firstStableIndex(nums []int, k int) int {
	N := len(nums)
	lpfx := make([]int, N)
	copy(lpfx, nums)
	for i := 0; i < N; i++ {
		if i == 0 {
			lpfx[i] = nums[i]
		} else {
			lpfx[i] = max(lpfx[i-1], nums[i])
		}
	}
	rpfx := make([]int, N)
	copy(rpfx, nums)
	for i := N - 1; i >= 0; i-- {
		if i == N-1 {
			rpfx[i] = nums[i]
		} else {
			rpfx[i] = min(rpfx[i+1], nums[i])
		}
	}
	for i := 0; i < N; i++ {
		if lpfx[i]-rpfx[i] <= k {
			return i
		}
	}
	return -1
}

func main() {
	type Pair struct {
		nums []int
		k    int
	}
	for _, v := range []Pair{Pair{[]int{5, 0, 1, 4}, 3},
		Pair{[]int{3, 2, 1}, 1},
		Pair{[]int{0}, 0},
	} {
		fmt.Printf("nums = %v ; k = %d\n", v.nums, v.k)
		r := firstStableIndex(v.nums, v.k)
		fmt.Printf("r = %d\n", r)
		fmt.Println("===================")
	}
}





